### ISSUE TITLE
Switching devices on the fly

### ISSUE THREAD

**Issue by cjolivier01** [pytorch/xla#2081]
## ❓ Questions and Help
Let’s say I want to run a model on a CPU or GPU for some time and then seamlessly switch to TPU without really leaving the training loop. obviously weights would need to be moved and the dataset stream redirected, and I can see from the code and behavior that it’s either possible or really close. Then the visa versa, from TPU back to CPU. 
options being create alternate tensors on the fly on another device in the training loop or around collecting the sync tensors? syncing tensor back to the host and then changing the device and continuing? redirect at XRT grpc layer? combination of those or something different?

**Comment by dlibenzi** [comment]
You will have to move move model, optimizer, ... states to the different devices, or checkpoint/restore them.
The issue is with dataset, especially when using things like the ParallelLoader, which does prefetching. And if you do not use the ParallelLoader, you lose quite a bit of perf, especially with input heavy models.
You will have to remember the dataset indices already consumed, and create custom datasets/samplers able to get a raw input dataset and a set of already consumed indices.
If you do not use the ParallelLoader life is easier from that POV.

**Comment by cjolivier01** [comment]
Thank you.  Point taken on ParallelLoader, it'll be ahead of me in samples and so there could be some missed if I don't put up some sort of "barrier" to pause it, catch up steps locally on CPU before transferring the model off-host to TPU.

So lets say I intercept the compile attempt in XLATensor::Compile() (the training thread is getting 100% cache hits for each step at this point, which is a requirement for my use-case) and recompile for TPU here.  Is there anything clever I can do to the existing live tensors in order to do one or more of the following?:

1. Invalidate the knowledge that weights (need-grad leafs with aliases, I assume) are already sync'd to a particular GPU/TPU such that downstream logic automatically sends them to another device (I assume I can sync them directly just before or after my compile in order to bring them local to the host before going back out again)
2. Have the XrtComputatioNClient::TransferToServer() seamlessly start sending new samples to the new destination.

The reason is that I am trying to do this transparently to the python layer, so I imagine this could be done by a full checkpoint and then restarting the training loop on a new device from python, but looking how to do this seamlessly on the back-end.  i.e. I wouldn't expect it to be as simple as just changing the Device of the XLATensor and then it all just does what it needs to do to make it happen, but also I have concerns about creating new XLATensors with the new device and doing an implicit graph operation that transfers them, because this seems like it would both detach the XLATensor from the proper python object (wouldn't it? maybe they can share the impl object?) as well as what becomes of the old XLATensors in this case is not clear to me.

I actually sort of have this working, but my way is kind of hacky and I feel like there could be some massaging that could "trick" the synchronization engine into doing what I want rather than side-banding it.

Does this makes sense?

**Comment by dlibenzi** [comment]
The notion of device has wider spread.
Each process has an assigned device, all the XLA tensors have devices, ...
You cannot have two processes handling the same device also.
Can this be done? This is software, so yes, pretty much anything can be done.
Does it make sense for you? Depends on how important is for you that use case, and whether a simpler alternative can be found.
Does it make sense for the project? I'd say the complexity-based ROI is close to zero 😉

**Comment by cjolivier01** [comment]
Thank you for the input!

**Comment by cjolivier01** [comment]
btw, have had this concept working for some time. basically can run on one device and after the graph coalesces, switches to a "proxy" device (one with slower compile, for instance).  non-training-threads can continue to use the non-proxy device while the main traiining thread executes on the proxy. no samples are lost as it clones the samples to the proxy device and frees them along with the standard device's resources. switching between devices at runtime, say, when the graph changes (end of epoch, breakpoint hit, etc) is smooth as well.  I agree it's probably not something incredibly useful for the project, but it is pretty cool and if you ever come across a general use-case, I'd be happy to clean it up.
